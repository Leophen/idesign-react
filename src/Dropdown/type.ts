export interface DropdownProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 下拉内容类名
   */
  contentClassName?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 下拉列表宽度
   */
  width?: React.CSSProperties["width"];
  /**
   * 列表项最大高度
   */
  maxHeight?: React.CSSProperties["height"];
  /**
   * 下拉项高度
   * @default medium
   */
  size?: "small" | "medium" | "large";
  /**
   * 下拉选中项
   */
  value?: string | number | Array<string | number>;
  /**
   * 下拉操作项
   * @default []
   */
  options?: Array<DropdownOption>;
  /**
   * 菜单触发位置
   * @default bottom
   */
  placement?:
  | 'top'
  | 'left'
  | 'right'
  | 'bottom'
  | 'top-left'
  | 'top-right'
  | 'bottom-left'
  | 'bottom-right'
  | 'left-top'
  | 'left-bottom'
  | 'right-top'
  | 'right-bottom';
  /**
   * 菜单触发方式
   * @default click
   */
  trigger?: 'hover' | 'click' | 'context-menu';
  /**
   * 是否可多选
   * @default false
   */
  multiple?: boolean;
  /**
   * 级联子层级展开方向
   * @default right
   */
  cascaderDirection?: 'left' | 'right';
  /**
   * 是否全局禁用
   * @default false
   */
  disabled?: boolean;
  /**
   * 点击菜单项触发事件
   */
  onClick?: (value: string | number | Array<string | number>, event?: React.MouseEvent) => void;
  /**
   * 切换下拉操作时触发
   */
  onTrigger?: (visible: boolean) => void;
}

export interface DropdownItemProps {
  /**
   * 下拉操作项内容
   * @default ''
   */
  content?: React.ReactNode;
  /**
   * 下拉操作项唯一标识
   */
  value: string | number;
  /**
   * 级联子项宽度
   */
  width?: React.CSSProperties["width"];
  /**
   * 级联子项最大高度
   */
  maxHeight?: React.CSSProperties["height"];
  /**
   * 单项是否禁用
   * @default false
   */
  disabled?: boolean;
  /**
   * 是否显示操作项之后的分隔线
   * @default false
   */
  divider?: boolean;
  /**
   * 该下拉项上方组标题
   */
  title?: string;
  /**
   * 点击时触发
   */
  onClick?: (dropdownItem: DropdownOption, event: React.MouseEvent) => void;
}

export interface DropdownMenuProps {
  /**
   * 下拉操作项
   * @default []
   */
  options?: Array<DropdownOption>;
  /**
   * 下拉列表宽度
   */
  width?: React.CSSProperties["width"];
  /**
   * 级联子项最大高度
   */
  maxHeight?: React.CSSProperties["height"];
  /**
   * 下拉项高度
   * @default medium
   */
  size?: "small" | "medium" | "large";
  /**
   * 级联子层级展开方向
   * @default right
   */
  cascaderDirection?: 'left' | 'right';
  /**
   * 是否可多选
   * @default false
   */
  multiple?: boolean;
  /**
   * 选中值
   */
  selectedValue?: string | number | Array<string | number>;
  /**
   * 点击菜单项触发事件
   */
  clickItem?: (dropdownItem: DropdownOption, event: React.MouseEvent) => void;
}

export type DropdownOption = { children?: Array<DropdownItemProps> } & DropdownItemProps & Record<string, any>;
