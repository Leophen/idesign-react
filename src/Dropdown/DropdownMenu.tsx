import React from 'react';
import classNames from 'classnames';
import _ from 'lodash';
import './index.scss';
import Icon from '../Icon';
import { DropdownMenuProps, DropdownOption } from './type';

const DropdownMenu: React.FC<DropdownMenuProps> = (props) => {
  const {
    options = [],
    width,
    maxHeight,
    size,
    cascaderDirection = 'right',
    multiple = false,
    selectedValue = [],
    clickItem = () => { }
  } = props

  const handleItemClick = (item: DropdownOption, event: React.MouseEvent) => {
    event.stopPropagation()
    !(item.children && item.children?.length > 0) && clickItem?.(item, event)
    item.onClick?.(item, event)
  }

  const clickCascaderItem = (item: DropdownOption, event: React.MouseEvent) => {
    clickItem?.(item, event)
    item.onClick?.(item, event)
  }

  return (
    <ul
      className={classNames(
        'i-dropdown__menu',
        cascaderDirection === 'left' && 'i-dropdown__menu-left'
      )}
      style={{ width: width ? width : 'auto', maxHeight, overflowY: maxHeight ? 'auto' : 'unset' }}
    >
      {options.map((item, index) => {
        return (
          <li key={`${item.value}${index}}`}>
            {item.title && <header className="i-dropdown__item-header">{item.title}</header>}
            <div
              className={classNames(
                'i-dropdown__item',
                size && `i-dropdown__item--size-${size}`,
                item.disabled && 'i-dropdown__item-is-disabled',
                item.divider && 'i-dropdown__item-has-divider',
                item.value === selectedValue && 'i-dropdown__item-is-active',
                item.children && item.children?.length > 0 && 'i-dropdown__item-cascader',
                multiple && 'i-dropdown__item-multiple'
              )}
              data-direction={cascaderDirection}
              data-disabled={item.disabled}
              onClick={!item.disabled ? ((e) => handleItemClick(item, e)) : () => { }}
            >
              {item.children && item.children?.length > 0 && cascaderDirection === 'left' && <Icon name="ArrowLeft" size={12} color="var(--i-font-2)" />}
              <div
                className={classNames(
                  'i-dropdown__item-txt',
                  ((!multiple && item.value === selectedValue) || (multiple && Array.isArray(selectedValue) && selectedValue.includes(item.value)))
                  && 'i-dropdown__item-txt-is-active'
                )}
              >
                {item.content}
              </div>
              {item.children && item.children?.length > 0 && cascaderDirection === 'right' && <Icon name="ArrowRight" size={12} color="var(--i-font-2)" />}
              {item.children && item.children?.length > 0 &&
                <DropdownMenu
                  options={item.children}
                  width={item.width}
                  maxHeight={item.maxHeight}
                  size={size}
                  cascaderDirection={cascaderDirection}
                  multiple={multiple}
                  selectedValue={selectedValue}
                  clickItem={!item.disabled ? clickCascaderItem : () => { }}
                />
              }
              {multiple &&
                <div className="i-dropdown__item-check">
                  {Array.isArray(selectedValue) && selectedValue.includes(item.value) &&
                    <Icon name="Check" size={12} />
                  }
                </div>
              }
            </div>
          </li>
        )
      })}
    </ul>
  )
}

export default DropdownMenu;
