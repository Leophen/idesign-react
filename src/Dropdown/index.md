---
nav:
  title: 组件
  path: /components
group:
  title: 导航组件
  order: 2
order: 0
---

# Dropdown 下拉菜单

基于指定节点的折叠下拉菜单。

## 基本用法

通过 `options` 属性指定下拉列表内容，并通过 `onClick` 指定点击下拉内容触发事件。

```tsx
import React from 'react'
import { Dropdown, Button } from 'idesign-react'

const App = () => {
  const options = [
    {
      content: '操作一',
      value: 1
    },
    {
      content: '操作二',
      value: 2
    },
    {
      content: '操作三',
      value: 3
    }
  ]

  const handleSelect = (val) => {
    console.log(val)
  }

  return (
    <Dropdown options={options} onClick={handleSelect}>
      <Button>更多</Button>
    </Dropdown>
  )
}

export default App
```

## 限制最大高度

可通过 `maxHeight` 属性指定下拉列表最大高度，注意最外层设最大高度后无法使用级联（待滚动条组件后解决）。

```tsx
import React from 'react'
import { Dropdown, Button } from 'idesign-react'

const App = () => {
  const options = Array(24).fill({
    content: '操作项',
    value: Math.random()
  })

  const handleSelect = (val) => {
    console.log(val)
  }

  return (
    <Dropdown options={options} maxHeight={300} onClick={handleSelect}>
      <Button>限制最大高度</Button>
    </Dropdown>
  )
}

export default App
```

## 全局禁用状态

可通过 `disabled` 属性控制禁用整个下拉菜单。

```tsx
import React from 'react'
import { Dropdown, Button } from 'idesign-react'

const App = () => {
  const options = [
    {
      content: '操作一',
      value: 1
    },
    {
      content: '操作二',
      value: 2
    },
    {
      content: '操作三',
      value: 3
    }
  ]

  const handleSelect = (val) => {
    console.log(val)
  }

  return (
    <Dropdown options={options} disabled={true} onClick={handleSelect}>
      <Button>禁用下拉菜单</Button>
    </Dropdown>
  )
}

export default App
```

## 单项禁用状态

也可在 `options` 属性中设置 `disabled` 控制单项禁用。

```tsx
import React from 'react'
import { Dropdown, Button } from 'idesign-react'

const App = () => {
  const options = [
    {
      content: '操作一',
      value: 1
    },
    {
      content: '操作二',
      value: 2,
      disabled: true
    },
    {
      content: '操作三',
      value: 3
    }
  ]

  const handleSelect = (val) => {
    console.log(val)
  }

  return (
    <Dropdown options={options} onClick={handleSelect}>
      <Button>禁用操作二</Button>
    </Dropdown>
  )
}

export default App
```

## 带分割线的下拉项

可在 `options` 属性中设置 `divider` 控制该项下方展示分割线。

```tsx
import React from 'react'
import { Dropdown, Button } from 'idesign-react'

const App = () => {
  const options = [
    {
      content: '操作一',
      value: 1
    },
    {
      content: '操作二',
      value: 2,
      divider: true
    },
    {
      content: '操作三',
      value: 3
    }
  ]

  const handleSelect = (val) => {
    console.log(val)
  }

  return (
    <Dropdown options={options} onClick={handleSelect}>
      <Button>带分割线的下拉菜单</Button>
    </Dropdown>
  )
}

export default App
```

## 单项点击事件

可在 `options` 属性中设置 `onClick` 控制该项点击触发事件。

```tsx
import React from 'react'
import { Dropdown, Button } from 'idesign-react'

const App = () => {
  const options = [
    {
      content: '操作一',
      value: 1
    },
    {
      content: '操作二',
      value: 2,
      onClick: (item, e) => { console.log(item, e) }
    },
    {
      content: '操作三',
      value: 3
    }
  ]

  return (
    <Dropdown options={options}>
      <Button>点击操作二触发</Button>
    </Dropdown>
  )
}

export default App
```

## 不同触发方式

可通过 `trigger` 属性指定下拉菜单触发方式，默认为 `click`。

```tsx
import React from 'react'
import { Dropdown, Button } from 'idesign-react'

const App = () => {
  const options = [
    {
      content: '操作一',
      value: 1
    },
    {
      content: '操作二',
      value: 2
    },
    {
      content: '操作三',
      value: 3
    }
  ]

  const handleSelect = (val) => {
    console.log(val)
  }

  return (
    <div className="idesign-demo-block-row">
      <Dropdown trigger="click" options={options} onClick={handleSelect}>
        <Button>点击触发</Button>
      </Dropdown>
      <Dropdown trigger="hover" options={options} onClick={handleSelect}>
        <Button>悬浮触发</Button>
      </Dropdown>
      <Dropdown trigger="context-menu" options={options} onClick={handleSelect}>
        <Button>右击触发</Button>
      </Dropdown>
    </div>
  )
}

export default App
```

## 可选中模式

可以通过指定 `value` 选中值将下拉菜单切换为可选中的模式。

```tsx
import React, { useState } from 'react'
import { Dropdown, Button } from 'idesign-react'

const App = () => {
  const options = [
    {
      content: '操作一',
      value: 1
    },
    {
      content: '操作二',
      value: 2
    },
    {
      content: '操作三',
      value: 3
    }
  ]

  const [value, setValue] = useState(2)
  const handleSelect = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <div className="idesign-demo-block-row">
      <Dropdown
        value={value}
        options={options}
        onClick={handleSelect}
      >
        <Button>可选中</Button>
      </Dropdown>
      <Dropdown
        value={value}
        options={options}
        onClick={handleSelect}
      >
        <Button>同步对照组</Button>
      </Dropdown>
    </div>
  )
}

export default App
```

## 下拉多选模式

可通过 `multiple` 属性指定下拉菜单为多选模式。

```tsx
import React, { useState } from 'react'
import { Dropdown, Button } from 'idesign-react'

const App = () => {
  const options = [
    {
      content: '操作一',
      value: 1
    },
    {
      content: '操作二',
      value: 2
    },
    {
      content: '操作三',
      value: 3
    }
  ]

  const [value, setValue] = useState([2])
  const handleSelect = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <div className="idesign-demo-block-row">
      <Dropdown
        value={value}
        options={options}
        multiple={true}
        onClick={handleSelect}
      >
        <Button>多选菜单</Button>
      </Dropdown>
      <Dropdown
        value={value}
        options={options}
        multiple={true}
        onClick={handleSelect}
      >
        <Button>同步对照组</Button>
      </Dropdown>
    </div>
  )
}

export default App
```

## 级联下拉菜单

可通过 `children` 属性指定级联菜单，可通过 `cascaderDirection` 控制展开方向。

```tsx
import React from 'react'
import { Dropdown, Button } from 'idesign-react'

const App = () => {
  const options = [
    {
      content: '操作 1',
      value: 1
    },
    {
      content: '操作 2',
      value: 2,
      children: [
        {
          content: '操作 2-1',
          value: 2.1
        },
        {
          content: '操作 2-2',
          value: 2.2,
          children: [
            {
              content: '操作 2-2-1',
              value: 2.21
            },
            {
              content: '操作 2-2-2',
              value: 2.22
            },
            {
              content: '操作 2-2-3',
              value: 2.23
            }
          ]
        },
        {
          content: '操作 2-3',
          value: 2.3
        }
      ]
    },
    {
      content: '操作 3',
      value: 3
    }
  ]

  const handleSelect = (val) => {
    console.log(val)
  }

  return (
    <div className="idesign-demo-block-row">
      <Dropdown options={options} onClick={handleSelect}>
        <Button>级联菜单</Button>
      </Dropdown>
      <Dropdown cascaderDirection="left" options={options} onClick={handleSelect}>
        <Button>向左展开的级联菜单</Button>
      </Dropdown>
    </div>
  )
}

export default App
```

## 限制级联子项最大高度

可通过 `maxHeight` 属性指定级联子项的最大高度，注意只能给最里层设最大高度。

```tsx
import React from 'react'
import { Dropdown, Button } from 'idesign-react'

const App = () => {
  const options = [
    {
      content: '操作 1',
      value: 1
    },
    {
      content: '操作 2',
      value: 2,
      maxHeight: 300,
      children: Array(24).fill({
        content: '操作项',
        value: Math.random()
      })
    },
    {
      content: '操作 3',
      value: 3
    }
  ]

  const handleSelect = (val) => {
    console.log(val)
  }

  return (
    <Dropdown options={options} onClick={handleSelect}>
      <Button>限制级联子项高度</Button>
    </Dropdown>
  )
}

export default App
```

## Dropdown API

<API hideTitle />

## DropdownItem API

| 属性 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| content | 下拉操作项内容 | `ReactNode` | `--` |
| value | 下拉操作项唯一标识 | `string〡number` | `--` |
| width | 宽度 | `number` | `--` |
| maxHeight | 最大高度 | `number` | `--` |
| disabled | 单项是否禁用 | `boolean` | `false` |
| divider | 是否显示操作项之后的分隔线 | `boolean` | `false` |
| onClick | 点击时触发 | `(dropdownItem: DropdownOption, event: React.MouseEvent) => void` | `--` |
