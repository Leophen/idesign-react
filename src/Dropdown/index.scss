.i-dropdown__menu {
  position: relative;
  box-sizing: border-box;
  margin: 0;
  padding: 5px 0;
  word-wrap: break-word;
  list-style: none;
  background-color: var(--i-bg);
  border: none;
  border-radius: 4px;
  outline: none;
}

.i-dropdown__item-header {
  margin: 0;
  padding: 5px 16px;
  color: var(--i-font-3);
  font-size: 12px;
  line-height: 22px;
  list-style: none;
  cursor: default;
}

.i-dropdown__item {
  position: relative;
  display: flex;
  align-items: center;
  box-sizing: border-box;
  margin: 0;
  padding: 5px 16px;
  color: var(--i-font-2);
  font-size: 14px;
  line-height: 22px;
  direction: ltr;
  white-space: nowrap;
  word-wrap: break-word;
  list-style: none;
  outline: none;
  cursor: pointer;
  transition: all 0.2s ease-out;
  &.i-dropdown__item--size-small {
    padding: 1px 16px;
    font-size: 12px;
  }
  &.i-dropdown__item--size-large {
    padding: 7px 16px;
  }
  &[data-direction='left'] {
    .icon-ArrowLeft {
      margin-right: 12px;
      margin-left: -8px;
    }
    .i-dropdown__item-txt {
      text-align: end;
    }
  }
  &[data-direction='right'] {
    .icon-ArrowRight {
      margin-right: -18px;
      margin-left: 12px;
    }
    .i-dropdown__item-txt {
      text-align: start;
    }
    &.i-dropdown__item-cascader {
      padding-right: 32px;
    }
  }
  .i-dropdown__menu {
    position: absolute;
    top: 0;
    left: calc(100% - 6px);
    z-index: 1;
    display: none;
    border: 1px solid var(--i-border);
    border-radius: 4px;
    box-shadow: var(--i-shadow);
    cursor: default;
    .i-dropdown__item-txt {
      white-space: nowrap;
    }
    &.i-dropdown__menu-left {
      right: calc(100% - 6px);
      left: unset;
    }
  }
  &:hover {
    background-color: var(--i-bg-active);
    & > .i-dropdown__menu {
      display: block;
    }
  }
  &[data-disabled] {
    .i-icon {
      color: var(--i-font-disabled) !important;
    }
    &:hover {
      & > .i-dropdown__menu {
        display: none;
      }
    }
  }
  &.i-dropdown__item-has-divider {
    margin-bottom: 11px;
    &::after {
      position: absolute;
      bottom: -6px;
      left: 0px;
      box-sizing: border-box;
      width: 100%;
      height: 1px;
      margin: 0;
      padding: 0;
      line-height: 0;
      list-style: none;
      background-color: var(--i-border);
      content: '';
    }
  }
  &.i-dropdown__item-is-active {
    background-color: var(--i-bg-active);
  }
  &.i-dropdown__item-is-disabled {
    color: var(--i-font-disabled);
    cursor: not-allowed;
    user-select: none;
    &:hover {
      background-color: unset;
    }
  }
  &.i-dropdown__item-multiple {
    padding-right: 32px;
    &.i-dropdown__item-is-active {
      background-color: unset;
    }
    &:hover {
      background-color: var(--i-bg-active);
    }
  }
}

.i-dropdown__item-txt {
  width: 100%;
  white-space: normal;
  &.i-dropdown__item-txt-is-active {
    color: var(--i-primary-txt);
  }
}

.i-dropdown__item-check {
  position: absolute;
  right: 12px;
  display: flex;
  align-items: center;
  width: 14px;
  height: 14px;
  .icon-Check {
    color: var(--i-primary-txt);
    pointer-events: none;
  }
}

@keyframes show {
  0% {
    display: none;
  }
  100% {
    display: block;
  }
}
