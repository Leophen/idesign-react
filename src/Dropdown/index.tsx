import React, { useEffect, useState } from 'react';
import classNames from 'classnames';
import _ from 'lodash';
import './index.scss';
import Popup from '../Popup';
import Icon from '../Icon';

export interface DropdownItemProps {
  /**
   * 下拉操作项内容
   * @default ''
   */
  content?: React.ReactNode;
  /**
   * 下拉操作项唯一标识
   */
  value?: string | number;
  /**
   * 级联子项宽度
   */
  width?: number;
  /**
   * 级联子项最大高度
   */
  maxHeight?: number;
  /**
   * 是否高亮当前操作项
   * @default false
   */
  active?: boolean;
  /**
   * 单项是否禁用
   * @default false
   */
  disabled?: boolean;
  /**
   * 是否显示操作项之后的分隔线
   * @default false
   */
  divider?: boolean;
  /**
   * 点击时触发
   */
  onClick?: (dropdownItem: DropdownOption, event: React.MouseEvent) => void;
}

export interface DropdownMenuProps {
  /**
   * 下拉操作项
   * @default []
   */
  options?: Array<DropdownOption>;
  /**
   * 下拉列表宽度
   */
  width?: number;
  /**
   * 级联子项最大高度
   */
  maxHeight?: number;
  /**
   * 级联子层级展开方向
   * @default right
   */
  cascaderDirection?: 'left' | 'right';
  /**
   * 是否可多选
   * @default false
   */
  multiple?: boolean;
  /**
   * 点击菜单项触发事件
   */
  clickItem?: (dropdownItem: DropdownOption, event: React.MouseEvent) => void;
}

export type DropdownOption = { children?: Array<DropdownItemProps> } & DropdownItemProps & Record<string, any>;

export interface DropdownProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 下拉列表宽度
   */
  width?: number;
  /**
   * 列表项最大高度
   */
  maxHeight?: number;
  /**
   * 下拉操作项
   * @default []
   */
  options?: Array<DropdownOption>;
  /**
   * 菜单触发位置
   * @default bottom
   */
  placement?:
  | 'top'
  | 'left'
  | 'right'
  | 'bottom'
  | 'top-left'
  | 'top-right'
  | 'bottom-left'
  | 'bottom-right'
  | 'left-top'
  | 'left-bottom'
  | 'right-top'
  | 'right-bottom';
  /**
   * 菜单触发方式
   * @default click
   */
  trigger?: 'hover' | 'click' | 'context-menu';
  /**
   * 是否可多选
   * @default false
   */
  multiple?: boolean;
  /**
   * 级联子层级展开方向
   * @default right
   */
  cascaderDirection?: 'left' | 'right';
  /**
   * 是否全局禁用
   * @default false
   */
  disabled?: boolean;
  /**
   * 点击菜单项触发事件
   */
  onClick?: (dropdownItem: DropdownOption, event: React.MouseEvent) => void;
  /**
   * 切换下拉操作时触发
   */
  onTrigger?: (visible: boolean) => void;
}

const DropdownMenu: React.FC<DropdownMenuProps> = (props) => {
  const {
    options = [],
    width,
    maxHeight,
    cascaderDirection = 'right',
    multiple = false,
    clickItem = () => { }
  } = props

  const handleItemClick = (item: DropdownOption, event: React.MouseEvent) => {
    !(item.children && item.children?.length > 0) && clickItem?.(item, event)
    item.onClick?.(item, event)
  }

  const clickCascaderItem = (item: DropdownOption, event: React.MouseEvent) => {
    clickItem?.(item, event)
    item.onClick?.(item, event)
  }

  return (
    <ul
      className={classNames(
        'i-dropdown__menu',
        cascaderDirection === 'left' && 'i-dropdown__menu-left'
      )}
      style={{ width: width ? width : 'auto', maxHeight, overflowY: maxHeight ? 'auto' : 'unset' }}
    >
      {options.map((item, index) => {
        return (
          <li
            className={classNames(
              'i-dropdown__item',
              item.disabled && 'i-dropdown__item-is-disabled',
              item.divider && 'i-dropdown__item-has-divider',
              item.active && 'i-dropdown__item-is-active',
              multiple && 'i-dropdown__item-multiple'
            )}
            data-direction={cascaderDirection}
            data-disabled={item.disabled}
            onClick={!item.disabled ? ((e) => handleItemClick(item, e)) : () => { }}
            key={`${item.value}${index}}`}
          >
            {item.children && item.children?.length > 0 && cascaderDirection === 'left' && <Icon name="ArrowLeft" size={12} color="rgba(0,0,0,.6)" />}
            <div className='i-dropdown__item-txt'>
              {item.content}
            </div>
            {item.children && item.children?.length > 0 && cascaderDirection === 'right' && <Icon name="ArrowRight" size={12} color="rgba(0,0,0,.6)" />}
            {item.children && item.children?.length > 0 &&
              <DropdownMenu
                options={item.children}
                width={item.width}
                maxHeight={item.maxHeight}
                cascaderDirection={cascaderDirection}
                multiple={multiple}
                clickItem={!item.disabled ? clickCascaderItem : () => { }}
              />
            }
            {multiple &&
              <div className="i-dropdown__item-check">
                {item.active && <Icon name="Check" size={14} />}
              </div>
            }
          </li>
        )
      })}
    </ul>
  )
}

const Dropdown: React.FC<DropdownProps> = (props) => {
  const {
    children = '',
    className,
    style,
    width,
    maxHeight,
    options = [],
    placement = 'bottom',
    trigger = 'click',
    cascaderDirection = 'right',
    multiple = false,
    disabled = false,
    onClick = () => { },
    onTrigger = () => { }
  } = props;

  const [popupVisible, setPopupVisible] = useState(false)

  const switchPopup = (visible: boolean) => {
    setPopupVisible(visible)
    onTrigger?.(visible)
  }

  const [innerOptions, setInnerOptions] = useState(options)
  useEffect(() => {
    setInnerOptions(options)
  }, [options])

  const [selectedList, setSelectedList] = useState<DropdownOption[]>([])
  const setItemStyle = (selectedArr: DropdownOption[]) => {
    const newOptions = _.cloneDeep(innerOptions)
    newOptions.map(item => {
      let itemSelected = false
      selectedArr.map(selectedItem => item.value === selectedItem.value && (itemSelected = true))
      item.active = itemSelected
    })
    setInnerOptions(newOptions)
  }
  const handleClickItem = (item: DropdownOption, event: React.MouseEvent) => {
    if (!multiple) {
      // 单选模式
      onClick?.(item, event)
      setPopupVisible(false)
      onTrigger?.(false)
    } else {
      // 多选模式
      const newSelectedList = _.cloneDeep(selectedList)
      let ifIncludeItem = false
      newSelectedList.map((it, index) => {
        if (it.value === item.value) {
          newSelectedList.splice(index, 1)
          ifIncludeItem = true
        }
      })
      !ifIncludeItem && newSelectedList.push(item)
      setSelectedList(newSelectedList)
      setItemStyle(newSelectedList)
      onClick?.(newSelectedList, event)
    }
  }

  const DropdownContent = (
    <div
      className={classNames(
        'i-dropdown',
        className
      )}
      style={{ ...(style || {}), ...{ width: width ? width : 'auto', maxHeight, overflowY: maxHeight ? 'auto' : 'unset' } }}
    >
      <DropdownMenu
        multiple={multiple}
        options={innerOptions}
        cascaderDirection={cascaderDirection}
        clickItem={handleClickItem}
      />
    </div>
  )

  return (
    <Popup
      content={DropdownContent}
      placement={placement}
      trigger={trigger}
      visible={popupVisible}
      disabled={disabled}
      onTrigger={switchPopup}
    >
      {children}
    </Popup>
  );
};

Dropdown.displayName = 'Dropdown';

export default Dropdown;
