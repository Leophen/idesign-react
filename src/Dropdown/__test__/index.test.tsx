import { render } from '@testing-library/react';
import Dropdown from '../index';

describe('Dropdown 组件测试', () => {
  const options = [
    {
      content: '操作一',
      value: 1
    },
    {
      content: '操作二',
      value: 2
    },
    {
      content: '操作三',
      value: 3
    }
  ]
  it('create', () => {
    const { asFragment } = render(
      <Dropdown options={options}>
        <button>更多</button>
      </Dropdown>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

