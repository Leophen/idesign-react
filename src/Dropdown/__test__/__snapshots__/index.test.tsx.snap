// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-popup__reference"
  >
    <button>
      更多
    </button>
  </div>
</DocumentFragment>
`;
